libclambcc/ClamBCOutlineEndiannessCalls/ClamBCOutlineEndiannessCalls.cpp

```diff
@@ -8,104 +8,136 @@
+                assert(pNew && "Could not create function");
 
-        BasicBlock* pEntry = BasicBlock::Create(pMod->getContext(), "entry", pNew, nullptr);
-        assert(pEntry && "Could not create basic block");
+                BasicBlock* pEntry = BasicBlock::Create(pMod->getContext(), "entry", pNew, nullptr);
+                assert(pEntry && "Could not create basic block");
 
-        CallInst* pCallInst = CallInst::Create(func, "newCall", pEntry);
-        assert(pCallInst && "Could not create call instruction");
+                CallInst* pCallInst = CallInst::Create(func, "newCall", pEntry);
+                assert(pCallInst && "Could not create call instruction");
 
-        ReturnInst* pReturnInst = ReturnInst::Create(pMod->getContext(), pCallInst, pEntry);
-        assert(pReturnInst && "Could not create return instruction.");
+                ReturnInst* pReturnInst = ReturnInst::Create(pMod->getContext(), pCallInst, pEntry);
+                assert(pReturnInst && "Could not create return instruction.");
 
-        pNew->addFnAttr(Attribute::OptimizeNone);
-        pNew->addFnAttr(Attribute::NoInline);
+                pNew->addFnAttr(Attribute::OptimizeNone);
+                pNew->addFnAttr(Attribute::NoInline);
 
-        //TODO: Test with NoInline, but not OptimizeNone (Hopefully I can have the function return 1 or 0, and
-        //not have to actually call the function.
+                //TODO: Test with NoInline, but not OptimizeNone (Hopefully I can have the function return 1 or 0, and
+                //not have to actually call the function.
 
-        return pNew;
-    }
+                return pNew;
+            }
 
-  public:
-    static char ID;
-    OutlineEndniassCalls()
-        : ModulePass(ID) {}
+        public:
+            OutlineEndiannessCalls()
+            {}
 
-    virtual bool runOnModule(Module& m) override
-    {
-        pMod = &m;
+            //virtual bool runOnModule(Module& m) override
+            virtual PreservedAnalyses run(Module & m, ModuleAnalysisManager & MAM)
+            {
+                pMod = &m;
 
-        std::vector<CallInst*> calls = findCalls();
+                std::vector<CallInst*> calls = findCalls();
 
-        if (0 == calls.size()) {
-            return false;
-        }
+                if (0 == calls.size()) {
+                    return PreservedAnalyses::all();
+                }
 
-        Function* pNew = getNewEndiannessFunction(calls[0]);
+                Function* pNew = getNewEndiannessFunction(calls[0]);
+
+                for (size_t i = 0; i < calls.size(); i++) {
+                    CallInst* pNewCall = CallInst::Create(pNew, "OutlineEndiannessCalls_", calls[i]);
+                    calls[i]->replaceAllUsesWith(pNewCall);
+                    calls[i]->eraseFromParent();
+                }
+
+                if (bChanged){
+                    return PreservedAnalyses::none();
+                }
+                return PreservedAnalyses::all();
```

Comment:
```suggestion
                if (bChanged) {
                    /* Since we changed the IR here invalidate all the previous analysis.
                     * We only want to invalidate the analysis when we change something,
                     * since it is expensive to compute.
                     */
                    return PreservedAnalyses::none();
                }
                /* We didn't change anything, so keep the previous analysis.*/
                return PreservedAnalyses::all();
```